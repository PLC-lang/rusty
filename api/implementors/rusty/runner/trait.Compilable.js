(function() {var implementors = {};
implementors["rusty"] = [];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()