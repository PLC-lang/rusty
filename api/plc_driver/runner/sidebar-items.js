window.SIDEBAR_ITEMS = {"fn":["compile","compile_and_run"],"struct":["MainType"]};