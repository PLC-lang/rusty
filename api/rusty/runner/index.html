<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `runner` mod in crate `rusty`."><meta name="keywords" content="rust, rustlang, rust-lang, runner"><title>rusty::runner - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../../dark.css" disabled><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><script defer src="../../main.js"></script>
    <noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../rusty/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../../rusty/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Module runner</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></div></section><div id="sidebar-vars" data-name="runner" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../rusty/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Module <a href="../index.html">rusty</a>::<wbr><a class="mod" href="#">runner</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/rusty/runner.rs.html#1-99">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.MainType.html" title="rusty::runner::MainType struct">MainType</a></div><div class="item-right docblock-short"></div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Compilable.html" title="rusty::runner::Compilable trait">Compilable</a></div><div class="item-right docblock-short"></div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.compile.html" title="rusty::runner::compile fn">compile</a></div><div class="item-right docblock-short"><p>Compiles and runs the given sources
Sources must be <code>Compilable</code>, default implementations include <code>String</code> and <code>&amp;str</code>
An implementation is also provided for <code>Vec&lt;SourceContainer&gt;</code></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.compile_and_run.html" title="rusty::runner::compile_and_run fn">compile_and_run</a></div><div class="item-right docblock-short"><p>A Convenience method to compile and then run the given source</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.run.html" title="rusty::runner::run fn">run</a></div><div class="item-right docblock-short"><p>Runs the function given by <code>name</code> inside the compiled execution engine code.
Returns the value returned by calling the function</p>
</div></div></div></section><section id="search" class="content hidden"></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="rusty" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.60.0 (7737e0b5c 2022-04-04)" ></div>
</body></html>