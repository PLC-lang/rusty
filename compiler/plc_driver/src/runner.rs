use crate::{pipelines::ParsedProject, CompileOptions};

use plc::codegen::{CodegenContext, GeneratedModule};
use plc_diagnostics::diagnostician::Diagnostician;
use plc_index::GlobalContext;
use project::project::Project;
use source_code::Compilable;

#[allow(dead_code)]
#[repr(C)]
pub struct MainType {
    a: [usize; 1000],
}

impl Default for MainType {
    fn default() -> Self {
        MainType { a: [0; 1000] }
    }
}

///
/// Compiles and runs the given sources
/// Sources must be `Compilable`, default implementations include `String` and `&str`
/// An implementation is also provided for `Vec<SourceContainer>`
///
pub fn compile<T: Compilable>(context: &CodegenContext, source: T) -> GeneratedModule<'_> {
    let source = source.containers();
    let project = Project::new("TestProject".to_string()).with_sources(source);
    let ctxt = GlobalContext::new().with_source(project.get_sources(), None).unwrap();
    let mut diagnostician = Diagnostician::null_diagnostician();
    let parsed_project = ParsedProject::parse(&ctxt, project, &mut diagnostician).unwrap();
    let indexed_project = parsed_project.index(ctxt.provider());
    let annotated_project = indexed_project.annotate(ctxt.provider());
    let compile_options = CompileOptions {
        optimization: plc::OptimizationLevel::None,
        debug_level: plc::DebugLevel::None,
        ..Default::default()
    };

    annotated_project.generate_single_module(context, &compile_options).unwrap().unwrap()
}

///
/// A Convenience method to compile and then run the given source
///
pub fn compile_and_run<T, U, S: Compilable>(source: S, params: &mut T) -> U {
    let context: CodegenContext = CodegenContext::create();
    let module = compile(&context, source);
    module.print_to_stderr();
    module.run::<T, U>("main", params)
}

///
/// A Convenience method to compile and then run the given source
/// without external parameters
///
pub fn compile_and_run_no_params<U, S: Compilable>(source: S) -> U {
    let context: CodegenContext = CodegenContext::create();
    let module = compile(&context, source);
    module.print_to_stderr();
    module.run_no_param::<U>("main")
}
