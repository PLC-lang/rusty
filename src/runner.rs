use inkwell::{
    context::Context,
    execution_engine::{ExecutionEngine, JitFunction},
};

use crate::{compile_module, diagnostics::Diagnostician, SourceCode, SourceContainer};

type MainFunction<T, U> = unsafe extern "C" fn(*mut T) -> U;
type MainEmptyFunction<U> = unsafe extern "C" fn() -> U;

#[allow(dead_code)]
#[repr(C)]
pub struct MainType {
    a: [usize; 1000],
}

impl Default for MainType {
    fn default() -> Self {
        MainType { a: [0; 1000] }
    }
}

pub trait Compilable {
    type T: SourceContainer;
    fn containers(self) -> Vec<Self::T>;
}

impl Compilable for &str {
    type T = SourceCode;
    fn containers(self) -> Vec<Self::T> {
        let code = Self::T::from(self);
        vec![code]
    }
}

impl Compilable for String {
    type T = SourceCode;
    fn containers(self) -> Vec<Self::T> {
        let code = self.into();
        vec![code]
    }
}

impl<S: SourceContainer> Compilable for Vec<S> {
    type T = S;
    fn containers(self) -> Vec<Self::T> {
        self
    }
}

impl Compilable for SourceCode {
    type T = Self;

    fn containers(self) -> Vec<Self::T> {
        vec![self]
    }
}

///
/// Runs the function given by `name` inside the compiled execution engine code.
/// Returns the value returned by calling the function
///
pub fn run<T, U>(exec_engine: &ExecutionEngine, name: &str, params: &mut T) -> U {
    unsafe {
        let main: JitFunction<MainFunction<T, U>> = exec_engine.get_function(name).unwrap();
        let main_t_ptr = &mut *params as *mut _;

        main.call(main_t_ptr)
    }
}

///
/// Runs the function given by `name` inside the compiled execution engine code.
/// Returns the value returned by calling the function
///
pub fn run_no_param<U>(exec_engine: &ExecutionEngine, name: &str) -> U {
    unsafe {
        let main: JitFunction<MainEmptyFunction<U>> = exec_engine.get_function(name).unwrap();
        main.call()
    }
}

///
/// Compiles and runs the given sources
/// Sources must be `Compilable`, default implementations include `String` and `&str`
/// An implementation is also provided for `Vec<SourceContainer>`
///
pub fn compile<T: Compilable>(context: &Context, source: T) -> ExecutionEngine {
    let source = source.containers();
    let (_, code_gen) = compile_module(
        context,
        source,
        vec![],
        None,
        Diagnostician::null_diagnostician(),
    )
    .unwrap();
    code_gen.module.print_to_stderr();
    code_gen
        .module
        .create_jit_execution_engine(inkwell::OptimizationLevel::None)
        .unwrap()
}

///
/// A Convenience method to compile and then run the given source
///
pub fn compile_and_run<T, U, S: Compilable>(source: S, params: &mut T) -> U {
    let context: Context = Context::create();
    let exec_engine = compile(&context, source);
    run::<T, U>(&exec_engine, "main", params)
}
