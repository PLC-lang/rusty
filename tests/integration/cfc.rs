use driver::runner::compile_and_run;

use crate::get_test_file;
mod resolver_tests;
mod validation_tests;

#[test]
fn variables_assigned() {
    // GIVEN a ST program with hardcoded values and CFC program with two variables but without a body
    let st_file = get_test_file("cfc/assigning.st");
    let cfc_file = get_test_file("cfc/assigning.cfc");

    // WHEN assigning these values to the CFC-POU variables and adding them together
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN we get the correct result
    assert_eq!(res, 300);
}

#[test]
fn simple_assignment() {
    // GIVEN a CFC program which assigns one variable to another
    let st_file = get_test_file("cfc/variable_assignment.st");
    let cfc_file = get_test_file("cfc/variable_assignment.cfc");
    // WHEN assigning values to them and then calling the program
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the second variable will have the value of the first variable
    assert_eq!(res, 10);
}

#[test]
fn select_call_in_function_block_with_input_variables() {
    // GIVEN a CFC program which selects a variable based on a predicate
    let st_file = get_test_file("cfc/select.st");
    let cfc_file = get_test_file("cfc/select.cfc");
    // WHEN assigning values to them and then calling the program
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the correct value is selected
    assert_eq!(res, 1);
}

#[test]
fn custom_function_call_in_function_block() {
    // GIVEN a CFC program which calls a subroutine
    let st_file = get_test_file("cfc/my_add.st");
    let cfc_file = get_test_file("cfc/my_add.cfc");
    // WHEN calling the
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the second variable will have the value of the first variable
    assert_eq!(res, 4);
}

#[test]
fn chained_calls() {
    // GIVEN a CFC program which assigns a variable
    let st_file = get_test_file("cfc/chained_calls.st");
    let cfc_file = get_test_file("cfc/chained_calls.cfc");
    // WHEN assigning values to them and then calling the program
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the second variable will have the value of the first variable
    assert_eq!(res, 10);
}

#[test]
fn chained_calls_galore() {
    // GIVEN a CFC program which assigns a variable
    let st_file = get_test_file("cfc/chained_calls_galore.st");
    let cfc_file = get_test_file("cfc/chained_calls_galore.cfc");
    // WHEN assigning values to them and then calling the program
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the second variable will have the value of the first variable
    assert_eq!(res, 88);
}

#[test]
fn function_returns() {
    // GIVEN a CFC function which doubles a value
    let st_file = get_test_file("cfc/function_returns.st");
    let cfc_file = get_test_file("cfc/function_returns.cfc");
    // WHEN passing a value into the function
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN it will return the correct value
    assert_eq!(res, 222);
}

#[test]
fn conditional_return_evaluating_true() {
    // GIVEN a CFC function which returns early if a given argument is 5 and
    // otherwise modifies the argument to be 10
    let st_file = get_test_file("cfc/conditional_return_evaluating_true.st");
    let cfc_file = get_test_file("cfc/conditional_return.cfc");

    // WHEN passing 5 as an argument
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN it will early return, leaving the argument unmodified (i.e. 5)
    assert_eq!(res, 5);
}

#[test]
fn conditional_return_evaluating_false() {
    // GIVEN a CFC function which returns early if a given argument is 5 and
    // otherwise modifies the argument to be 10
    let st_file = get_test_file("cfc/conditional_return_evaluating_false.st");
    let cfc_file = get_test_file("cfc/conditional_return.cfc");

    // WHEN passing 0 as an argument
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN it will NOT return early and modify the argument to be 10
    assert_eq!(res, 10);
}

#[test]
fn conditional_return_evaluating_true_negated() {
    // GIVEN a CFC function which returns early if a given argument is NOT 5 and
    // otherwise modifies the argument to be 10
    let st_file = get_test_file("cfc/conditional_return_evaluating_true_negated.st");
    let cfc_file = get_test_file("cfc/conditional_return_negated.cfc");

    // WHEN passing 5 as an argument
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN it will NOT return early, modifying the argument to be 10
    assert_eq!(res, 10);
}

#[test]
fn conditional_return_block_evaluating_true() {
    // GIVEN a CFC function which returns early if variable argument `a` is bigger than `b` and otherwise
    // modifies an argument `res` to be 10
    let st_file = get_test_file("cfc/conditional_return_block_evaluating_true.st");
    let cfc_file = get_test_file("cfc/conditional_return_block.cfc");

    // WHEN passing variables a = 1 and b = 0
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN it will return early, leaving `res` unmodified
    assert_eq!(res, 5);
}

#[test]
fn conditional_return_block_evaluating_false() {
    // GIVEN a CFC function which returns early if variable argument `a` is bigger than `b` and otherwise
    // modifies an argument `res` to be 10
    let st_file = get_test_file("cfc/conditional_return_block_evaluating_false.st");
    let cfc_file = get_test_file("cfc/conditional_return_block.cfc");

    // WHEN passing variables a = 0 and b = 1
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});

    // THEN it will NOT return early, modifying `res` to be 10
    assert_eq!(res, 10);
}

#[test]
fn connection_sink_source() {
    // GIVEN a CFC program which assigns variables through a sink-source-pair and adds them together
    let st_file = get_test_file("cfc/connection.st");
    let cfc_file = get_test_file("cfc/connection_var_source_multi_sink.cfc");
    // WHEN calling the program
    let res: i32 = compile_and_run(vec![st_file, cfc_file], &mut {});
    // THEN the result will have double the value of the initial value
    assert_eq!(res, 4);
}

#[test]
fn jump_to_label_with_true() {
    let cfc_file = get_test_file("cfc/jump_true.cfc");

    let res: i32 = compile_and_run(vec![cfc_file], &mut {});
    assert_eq!(res, 3);
}

#[test]
fn jump_to_label_with_false() {
    let cfc_file = get_test_file("cfc/jump_false.cfc");

    let res: i32 = compile_and_run(vec![cfc_file], &mut {});
    assert_eq!(res, 5);
}

#[test]
fn actions_called_correctly() {
    #[derive(Default)]
    #[repr(C)]
    struct MainType {
        a: i32,
        b: i32,
    }

    let mut main = MainType::default();

    let file = get_test_file("cfc/actions.cfc");
    let _: i32 = compile_and_run(vec![file], &mut main);

    assert_eq!(main.a, 1);
    assert_eq!(main.b, 2);
}

// TODO(volsa): Remove this once our `test_utils.rs` file has been polished to also support CFC.
// More specifically transform the following tests into simple codegen ones.
#[cfg(test)]
mod ir {
    use driver::{compile, generate_to_string, generate_to_string_debug};
    use insta::assert_snapshot;
    use plc_source::SourceCode;
    use plc_xml::serializer::{
        SAction, SBlock, SConnector, SContinuation, SInVariable, SJump, SLabel, SOutVariable, SPou, SReturn,
    };

    use crate::get_test_file;

    const NEWLINE: &str = if cfg!(windows) { "\r\n" } else { "\n" };

    #[test]
    fn conditional_return() {
        let cfc_file = get_test_file("cfc/conditional_return.cfc");

        let output_file = tempfile::NamedTempFile::new().unwrap();
        let output_file_path = output_file.path().to_string_lossy();
        compile(&["plc", &cfc_file, "--ir", "-o", &output_file_path]).unwrap();
        let res = generate_to_string("plc", vec![cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    fn conditional_return_evaluating_true() {
        let st_file = get_test_file("cfc/conditional_return_evaluating_true.st");
        let cfc_file = get_test_file("cfc/conditional_return.cfc");

        let res = generate_to_string("plc", vec![st_file, cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    fn conditional_return_evaluating_true_negated() {
        let st_file = get_test_file("cfc/conditional_return_evaluating_true_negated.st");
        let cfc_file = get_test_file("cfc/conditional_return_negated.cfc");

        let res = generate_to_string("plc", vec![st_file, cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    fn variable_source_to_variable_and_block_sink() {
        let st_file = get_test_file("cfc/connection.st");
        let cfc_file = get_test_file("cfc/connection_var_source_multi_sink.cfc");

        let res = generate_to_string("plc", vec![st_file, cfc_file]).unwrap();
        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    #[ignore = "block-to-block connections not yet implemented"]
    fn block_source_to_variable_and_block_sink() {
        let st_file = get_test_file("cfc/connection.st");
        let cfc_file = get_test_file("cfc/connection_block_source_multi_sink.cfc");

        let res = generate_to_string("plc", vec![st_file, cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    fn jump_to_label_with_true() {
        let cfc_file = get_test_file("cfc/jump_true.cfc");

        let res = generate_to_string("plc", vec![cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    fn jump_to_label_with_false() {
        let cfc_file = get_test_file("cfc/jump_false.cfc");

        let res = generate_to_string("plc", vec![cfc_file]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    // TODO: Transfer this test to `codegen/tests/debug_tests/cfc.rs` once `test_utils.rs` has been refactored
    fn conditional_return_debug() {
        let declaration = "FUNCTION foo VAR_INPUT val : DINT; END_VAR";
        let content = SPou::init("foo", "function", declaration).with_fbd(vec![
            // IF val = 1 THEN RETURN
            &SInVariable::id(2).with_expression("val = 5"),
            &SReturn::id(3).with_execution_id(2).connect(2).negate(false),
            // ELSE val := 10
            &SInVariable::id(4).with_expression("10"),
            &SInVariable::id(5).with_execution_id(3).connect(4).with_expression("val"),
        ]);

        let mut source = SourceCode::from(content.serialize());
        source.path = Some("<internal>.cfc".into());
        let res = generate_to_string_debug("plc", vec![source]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);

        // We expect two different !dbg statements for the return statement and its condition
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    // TODO: Transfer this test to `codegen/tests/debug_tests/cfc.rs` once `test_utils.rs` has been refactored
    fn jump_debug() {
        let declaration = "PROGRAM foo VAR val : DINT := 0; END_VAR";
        let content = SPou::init("foo", "program", declaration).with_fbd(vec![
            // IF TRUE THEN GOTO lbl
            &SInVariable::id(1).with_expression("val = 0"), // condition
            &SLabel::id(2).with_name("lbl").with_execution_id(1), // label
            &SJump::id(3).with_name("lbl").with_execution_id(2).connect(1), // statement
            // ELSE x := FALSE
            &SOutVariable::id(4).with_execution_id(3).with_expression("val").connect(5), // assignment
            &SInVariable::id(5).with_expression("1"),
        ]);

        let mut source = SourceCode::from(content.serialize());
        source.path = Some("<internal>.cfc".into());
        let res = generate_to_string_debug("plc", vec![source]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);

        // We expect four different !dbg statement for the condition, label, statement and the assignment
        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    // TODO: Transfer this test to `codegen/tests/debug_tests/cfc.rs` once `test_utils.rs` has been refactored
    fn actions_debug() {
        let content = SPou::init("main", "program", "PROGRAM main VAR a, b : DINT; END_VAR")
            .with_actions(vec![
                &SAction::name("newAction").with_fbd(vec![
                    &SOutVariable::id(1).with_expression("a").with_execution_id(0).connect(2),
                    &SInVariable::id(2).with_expression("a + 1"),
                ]),
                &SAction::name("newAction2").with_fbd(vec![
                    &SInVariable::id(1).with_expression("b + 2"),
                    &SOutVariable::id(2).with_expression("b").with_execution_id(0).connect(1),
                ]),
            ])
            .with_fbd(vec![
                &SBlock::id(1).with_name("newAction").with_execution_id(1),
                &SBlock::id(2).with_name("newAction2").with_execution_id(2),
                &SInVariable::id(4).with_expression("0"),
                &SOutVariable::id(3).with_expression("a").with_execution_id(0).connect(4),
            ]);

        let mut source = SourceCode::from(content.serialize());
        source.path = Some("<internal>.cfc".into());
        let res = generate_to_string_debug("plc", vec![source]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);

        assert_snapshot!(output_file_content_without_headers);
    }

    #[test]
    // TODO: Transfer this test to `codegen/tests/debug_tests/cfc.rs` once `test_utils.rs` has been refactored
    fn sink_source_debug() {
        let content = SPou::init("main", "program", "PROGRAM main VAR x: DINT; END_VAR").with_fbd(vec![
            &SInVariable::id(1).with_expression("5"),
            &SConnector::id(2).with_name("s1").connect(1),
            &SContinuation::id(3).with_name("s1"),
            &SOutVariable::id(4).with_expression("x").with_execution_id(1).connect(3),
        ]);

        let mut source = SourceCode::from(content.serialize());
        source.path = Some("<internal>.cfc".into());
        let res = generate_to_string_debug("plc", vec![source]).unwrap();

        // We truncate the first 3 lines of the snapshot file because they contain file-metadata that changes
        // with each run. This is due to working with temporary files (i.e. tempfile::NamedTempFile::new())
        let output_file_content_without_headers = res.lines().skip(3).collect::<Vec<&str>>().join(NEWLINE);

        assert_snapshot!(output_file_content_without_headers);
    }
}
