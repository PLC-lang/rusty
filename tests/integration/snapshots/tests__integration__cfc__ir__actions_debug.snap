---
source: tests/integration/cfc.rs
expression: output_file_content_without_headers
---
%main = type { i32, i32 }

@main_instance = global %main zeroinitializer, section "var-$RUSTY$main_instance:r2i32i32", !dbg !0

define void @main(%main* %0) section "fn-$RUSTY$main:v" !dbg !12 {
entry:
  call void @llvm.dbg.declare(metadata %main* %0, metadata !16, metadata !DIExpression()), !dbg !17
  %a = getelementptr inbounds %main, %main* %0, i32 0, i32 0, !dbg !18
  %b = getelementptr inbounds %main, %main* %0, i32 0, i32 1, !dbg !18
  store i32 0, i32* %a, align 4, !dbg !19
  call void @main.newAction(%main* %0), !dbg !17
  call void @main.newAction2(%main* %0), !dbg !20
  ret void, !dbg !20
}

define void @main.newAction(%main* %0) section "fn-$RUSTY$main.newAction:v" !dbg !21 {
entry:
  call void @llvm.dbg.declare(metadata %main* %0, metadata !16, metadata !DIExpression()), !dbg !22
  %a = getelementptr inbounds %main, %main* %0, i32 0, i32 0, !dbg !23
  %b = getelementptr inbounds %main, %main* %0, i32 0, i32 1, !dbg !23
  %load_a = load i32, i32* %a, align 4, !dbg !22
  %tmpVar = add i32 %load_a, 1, !dbg !22
  store i32 %tmpVar, i32* %a, align 4, !dbg !22
  ret void, !dbg !22
}

define void @main.newAction2(%main* %0) section "fn-$RUSTY$main.newAction2:v" !dbg !24 {
entry:
  call void @llvm.dbg.declare(metadata %main* %0, metadata !16, metadata !DIExpression()), !dbg !25
  %a = getelementptr inbounds %main, %main* %0, i32 0, i32 0, !dbg !26
  %b = getelementptr inbounds %main, %main* %0, i32 0, i32 1, !dbg !26
  %load_b = load i32, i32* %b, align 4, !dbg !27
  %tmpVar = add i32 %load_b, 2, !dbg !27
  store i32 %tmpVar, i32* %b, align 4, !dbg !27
  ret void, !dbg !27
}

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare void @llvm.dbg.declare(metadata, metadata, metadata) #0

attributes #0 = { nofree nosync nounwind readnone speculatable willreturn }

!llvm.module.flags = !{!8, !9}
!llvm.dbg.cu = !{!10}

!0 = !DIGlobalVariableExpression(var: !1, expr: !DIExpression())
!1 = distinct !DIGlobalVariable(name: "main", scope: !2, file: !2, line: 1, type: !3, isLocal: false, isDefinition: true)
!2 = !DIFile(filename: "<internal>.cfc", directory: "")
!3 = !DICompositeType(tag: DW_TAG_structure_type, name: "main", scope: !2, file: !2, line: 1, size: 64, align: 64, flags: DIFlagPublic, elements: !4, identifier: "main")
!4 = !{!5, !7}
!5 = !DIDerivedType(tag: DW_TAG_member, name: "a", scope: !2, file: !2, line: 1, baseType: !6, size: 32, align: 32, flags: DIFlagPublic)
!6 = !DIBasicType(name: "DINT", size: 32, encoding: DW_ATE_signed, flags: DIFlagPublic)
!7 = !DIDerivedType(tag: DW_TAG_member, name: "b", scope: !2, file: !2, line: 1, baseType: !6, size: 32, align: 32, offset: 32, flags: DIFlagPublic)
!8 = !{i32 2, !"Dwarf Version", i32 5}
!9 = !{i32 2, !"Debug Info Version", i32 3}
!10 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "RuSTy Structured text Compiler", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, globals: !11, splitDebugInlining: false)
!11 = !{!0}
!12 = distinct !DISubprogram(name: "main", linkageName: "main", scope: !2, file: !2, line: 1, type: !13, scopeLine: 1, flags: DIFlagPublic, spFlags: DISPFlagDefinition, unit: !10, retainedNodes: !15)
!13 = !DISubroutineType(flags: DIFlagPublic, types: !14)
!14 = !{null}
!15 = !{}
!16 = !DILocalVariable(name: "main", scope: !12, file: !2, line: 1, type: !3)
!17 = !DILocation(line: 1, scope: !12)
!18 = !DILocation(line: 0, scope: !12)
!19 = !DILocation(line: 3, scope: !12)
!20 = !DILocation(line: 2, scope: !12)
!21 = distinct !DISubprogram(name: "main.newAction", linkageName: "main.newAction", scope: !2, file: !2, type: !13, scopeLine: 1, flags: DIFlagPublic, spFlags: DISPFlagDefinition, unit: !10, retainedNodes: !15)
!22 = !DILocation(line: 1, scope: !21)
!23 = !DILocation(line: 0, scope: !21)
!24 = distinct !DISubprogram(name: "main.newAction2", linkageName: "main.newAction2", scope: !2, file: !2, type: !13, scopeLine: 1, flags: DIFlagPublic, spFlags: DISPFlagDefinition, unit: !10, retainedNodes: !15)
!25 = !DILocation(line: 1, scope: !24)
!26 = !DILocation(line: 0, scope: !24)
!27 = !DILocation(line: 2, scope: !24)
